### ISSUE TITLE
Soundxyz transformer must take encoded tokenURI parameter and parse to URI

### ISSUE THREAD

**Issue by TimDaub** [neume-network/strategies#39]
- this file: https://github.com/neume-network/strategies/blob/main/src/strategies/soundxyz/transformer.mjs
- but, the extraction step only returns a hexadecimal value from the ethereum call
- this is how the transformer should look: https://github.com/neume-network/strategies/blob/a0a32a35d43666ccd73c45670f767573f253e996/src/strategies/soundxyz/transformer.mjs
- downloading the https tokenURI happens in another strategy

**Comment by il3ven** [comment]
I was looking into solving this. Essentially the flow is not clear to me. Can you give an overview of the application? Even after reading the code I am a little confused. The flow chart in Readme depicts a linear flow but the code in lifecycle.mjs paints a picture of a loop.

Also, how can I run the code to get a better understanding?

----

Regarding this issue: The issue's title says to parse the encoded tokenURI to URI. However, haven't you already done that here? Based upon my assumptions and understanding the output of `decodeCallOutput(["string"], line)` should be a simple string such as `https://sound.xyz/api/metadata/66/1/19`.

https://github.com/neume-network/strategies/blob/a0a32a35d43666ccd73c45670f767573f253e996/src/strategies/soundxyz/transformer.mjs#L8

**Comment by TimDaub** [comment]
![image](https://user-images.githubusercontent.com/2758453/169693004-1e0f9c4f-fd55-4f7b-869e-4deb61be48de.png)

I'm assuming you mean this chart above. Here "one `update` cycle later..." is supposed to indicate that indeed as you correctly pointed out, an extractor component has a loop flow.

Here's a pseudo-code of the lifecycle 

```
strategy = initByName("strategy1").
state0 = {};

{messages1, state as stateN, write as write1 } = strategy.init(state0);
if messages1.length == 0: return;
messages1.forEach(dispatch)

on(message, messageN => {
  {messagesN, state as stateN, write as writeN} = strategy.update(messageN, stateN)
  if messagesN.length == 0: return;
  messagesN.forEach(dispatch)
})
```

It confirms your observation, that if `messagesN` isn't an empty list, a loop can be created that permanently fires `strategy.update`.

Please let me know if this makes sense. I'm happy to provide more information.

> Regarding this issue: The issue's title says to parse the encoded tokenURI to URI. However, haven't you already done that here? Based on my assumptions and understanding, the output of `decodeCallOutput(["string"], line)` should be a simple string such as `https://sound.xyz/api/metadata/66/1/19`.

Yeah so here's the context of this issue. I made a mistake in that I created [a transformer that takes a line of JSON metadata and transforms it to a normalized neume-network/schema](https://github.com/neume-network/strategies/blob/main/src/strategies/soundxyz/transformer.mjs). Instead, since we're only querying the `tokenURI` from the contract, the json-rpc message will return a hexadecimal formatted value `0x....`. It's because we haven't called `decodeCallOutput` on it yet.

So the issue is about (re)moving the current code in `soundxyz/transformer.mjs` and replacing it with code that hasn't made it into the main branch yet. I guess that if you just copied that code and made a PR, the issue would be resolved: https://github.com/neume-network/strategies/blob/a0a32a35d43666ccd73c45670f767573f253e996/src/strategies/soundxyz/transformer.mjs And yes, the output should be e.g. "https://sound.xyz/api/metadata/66/1/19"

> Also, how can I run the code to get a better understanding?

- Install neume-network/core: https://github.com/neume-network/core#installation
- Make sure to set the correct env variables. If u end up needing a RPC_API_KEY, please let me know
- Please be conscious that after the renaming of the npm package names may not work out yet, e.g. `@neume-network/extraction-worker` must be replaced with `@music-os/extraction-worker` and so on.
- `npm run dev`. It'll first launch the "web3subgraph" strategy and then begin requesting soundxyz URIs

**Comment by il3ven** [comment]
Thank you for the explanation. The pseudo-code clears things up. Currently I am not able to visualise the complete application. 
- What are we supposed to use the state for? 
- Init makes sense as it creates the messages to fetch NFTs. What is the thinking behind update? 
- Why do we need multiple update cycles? 
- Are transformations applied after extraction is complete? Is transformation the last step?

I think all this will make more sense once I see a working example. Also, solving small issues like this one helps in building an understanding.

For now, I will create a PR for this issue. Earlier, I thought I was missing something. Why would you open an issue for copying code from one branch to another.

**Comment by TimDaub** [comment]
> * What are we supposed to use the state for?

I had initially planned to use state to keep track of e.g. variables like web3subgraph's "[lastId](https://github.com/neume-network/strategies/blob/6321b83e4cc1ddb1b2551f1137e103fae356fb33/src/strategies/web3subgraph/extractor.mjs#L15)", but now that any function call can dispatch a number of messages, it seems it's not required anymore. We might be able to remove it again.

> * Init makes sense as it creates the messages to fetch NFTs. What is the thinking behind update?

`update` is used to handle the messages' responses.

> * Why do we need multiple update cycles?

E.g. in web3subgraph, the `lastId` always just becomes apparent after having queried the graphql endpoint once.

> * Are transformations applied after extraction is complete? Is transformation the last step?

Actually, in the current latest "main" branch, an extractor must now manually apply transformations as can e.g. be seen here: https://github.com/neume-network/strategies/blob/6321b83e4cc1ddb1b2551f1137e103fae356fb33/src/strategies/web3subgraph/extractor.mjs#L51

I ended up fixing the issue myself now as I just spent a few hours trying to untangle what I had initially designed as the lifecycle. If you look into that file now, things should be a bit clearer I hope. Sorry for the confusing code so far...
